src/tools/rbd/Utils.cc

```diff
@@ -815,6 +815,31 @@ int snap_set(librbd::Image &image, const std::string &snap_name) {
   return 0;
 }
 
+bool calc_sparse_extent(const bufferptr &bp, size_t sparse_size, uint64_t length,
+                       size_t &write_offset, size_t &write_length,
+                       size_t &pos, size_t &offset) {
+  size_t shift;
+  if (pos + sparse_size > length) {
+    shift = length - pos;
+  } else {
+    shift = sparse_size;
+  }
+
+  bufferptr page(bp, offset, shift);
+  offset += shift;
+  pos += shift;
+
+  bool page_is_zero = page.is_zero();
+  if (!page_is_zero) {
+    write_length += shift;
+  }
+
+  if ((page_is_zero && write_length != 0) || pos == length) {
```

Comment:
Sorry for the typo in my last comment, I mean "if the page_is_zero is false the write_length is always 0". Then it should be that:
```
/* Do a write for this extent when it is not zero or it is the last extent. */
if (!page_is_zero || pos == length) {
  return true;
}
```
Am I understanding it correctly? thanx